### PR TITLE
improve and completely rework exclude optimizer

### LINKED ISSUE #79: common path algorithm calculates parts wrong with `/` inside part
Given a nested file structure:
<details>
  <summary>Click me to open file tree</summary>

```tree
.
├── a
│   ├── e
│   │   ├── h
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── i
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── j
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── n.txt
│   │   ├── o.txt
│   │   └── p.txt
│   ├── f
│   │   ├── h
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── i
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── j
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── n.txt
│   │   ├── o.txt
│   │   └── p.txt
│   ├── g
│   │   ├── h
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── i
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── j
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── n.txt
│   │   ├── o.txt
│   │   └── p.txt
│   ├── q.txt
│   ├── r.txt
│   └── s.txt
├── b
│   ├── e
│   │   ├── h
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── i
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── j
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── n.txt
│   │   ├── o.txt
│   │   └── p.txt
│   ├── f
│   │   ├── h
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── i
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── j
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── n.txt
│   │   ├── o.txt
│   │   └── p.txt
│   ├── g
│   │   ├── h
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── i
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── j
│   │   │   ├── k.txt
│   │   │   ├── l.txt
│   │   │   └── m.txt
│   │   ├── n.txt
│   │   ├── o.txt
│   │   └── p.txt
│   ├── q.txt
│   ├── r.txt
│   └── s.txt
```

</details>

When running the glob `./{a/e,b/f}/{n,o,k}.txt`, I expect to match the files `./a/e/n.txt`, `./a/e/o.txt`, `./b/f/n.txt`, `./b/f/o.txt`. `tinyglobby` however does not match any files.
When running the glob `./{a/e,b/f,**/*}/{n,o,k}.txt`, I expect to match a total of 30 files. `tinyglobby` however does not match any files.

I've created a minimal reproduction example in https://github.com/jxn-30/tinyglobby-brace-expansion that shows how `tinyglobby` differs from `fast-glob`/`globby` at this point.
[Glob tester](https://globster.xyz/) also shows how `minimatch` produces the same result as `fast-glob`/`globby`.

### PR DESCRIPTION
**Pull request by SuperchupuDev**
shorter and probably faster implementation :tada: also improves optimization for patterns that end with `**`

~~needs some work, this implementation (and the previous one) don't work well with patterns that wrap `/` in things like `*(src/index.ts)`. see https://github.com/SuperchupuDev/tinyglobby/pull/76#issuecomment-2599331217~~ disabled optimizations in those cases for now

also closes #79 once the above is solved, and makes tests from #80 work :-)

EDIT: REWRITTEN FROM THE GROUND UP with a different approach to avoid the many edge cases it used to have

### CODE REVIEW

**Comment by SuperchupuDev** [comment]
while the tests do pass now thanks to the new fdir version fixing excluding symlinks, this implementation doesn't handle patterns that have `/` inside glob symbols well, for example this test that i haven't pushed:
```ts
test('dynamic patterns that include slashes inside parts', async () => {
  const files = await glob({ patterns: ['{.a/a,a}/a.txt'], cwd });
  // files ends up being []
  assert.deepEqual(files.sort(), ['a/a.txt', '.a/a/a.txt']);
});
```

it's easy to see why it fails but hard to properly fix. other than splitting the parts with picomatch which is easy to do, i'm not sure how to modify the code to properly account for it. there's an option to just replace parts that have `/` with `**` but it should be the last choice as that would just skip optimizations

**Comment by benmccann** [comment]
Would it help to have `tinyglobby` expand `['{.a/a,a}/a.txt']` to `['.a/a/a.txt', 'a/a.txt']` as a first step?

**Comment by SuperchupuDev** [comment]
maybe but i'm not sure if it's optimal

**Comment by SuperchupuDev** [comment]
globbing update: this PR breaks some use cases of broken parentheses, because for some weird reason globbing unescaped parentheses also matches literal parentheses. i really wish we could get a partial option on picomatch

**Comment by SuperchupuDev** [comment]
the optimizer was completely rewritten again! the above comment no longer applies. should be ready to go :-) will test some more

File: src/utils.ts [hunk 1]
```diff
@@ -1,4 +1,64 @@
...
+        const match = regex[j].test(inputParts[j]);
+
+        if (!match) {
+          break;
+        }
+
+        // unlike popular belief, `**` doesn't return true in *all* cases
+        // some examples are when matching it to `.a` with dot: false or `..`
+        // so it needs to match to return early
+        if (part === '**' && match) {
```
Comment (Fuzzyma) [inline src/utils.ts:39]: You checked for a match in the if above and break out the loop. So match should be always truthy here. Hence it can be removed

**Review by Fuzzyma — COMMENTED** [review]
Found a little nitpick :D.

Also a "question" regarding your use of regexes. I learned, that regexes need to be compiled. So creating them via `/regex/` on the fly might double up work. If the regexes are likely to be used anyway (and more than once), define them on the module level so the compile step is only done once

File: src/utils.ts [hunk 2]
```diff
@@ -1,4 +1,63 @@
...
+}
+
+// the result of over 4 months of figuring stuff out and a LOT of help
+export function getPartialMatcher(patterns: string[], options?: PartialMatcherOptions): Matcher {
+  const regexes = patterns.map(pattern => splitPattern(pattern).map(part => picomatch.makeRe(part, options)));
+  return (input: string) => {
+    // no need to `splitPattern` as this is indeed not a pattern
+    const inputParts = input.split('/');
+    for (let i = 0; i < patterns.length; i++) {
+      const patternParts = splitPattern(patterns[i]);
```
Comment (pralkarz) [inline src/utils.ts:16]: Aren't we duplicating some work on lines 11 (`patterns.map(pattern => splitPattern(pattern)...`) and 16?

Comment (SuperchupuDev) [reply src/utils.ts:16]: looks like we are, will fix

**Comment by SuperchupuDev** [comment]
@Fuzzyma @pralkarz suggestions have been applied in 8872e68, let me know if y'all want me to change anything else :+1:
